### DISCUSSION TITLE
Improved preview options

### DISCUSSION THREAD

**Discussion by WoLfulus** [directus/directus#19044] [Feature Requests]
### Summary

Would be nice to be able to provide additional interface elements (input, dropdowns, and things like this) to the preview window, or at least a way to implement extensions on top of it to add custom interface elements like these (+ being able to dispatch messages to the preview iframe):

![image](https://github.com/directus/directus/assets/412397/a0069c15-7880-426b-8168-5a7ca886e216)

### Basic Example

It could be either a repeater where you can select the type of interface, or an extension where you define what needs to be rendered on the preview top bar.

```
export default definePreviewElement({
  id: "language-selection",
  component: MyDropdownLanguage,
  options: ...
});
```

### Motivation

One use case I have right now is that I have to switch between languages in the preview mode (which changes the URL). I had to write a wrapping page + another iframe in order to switch between languages since what I'm previewing isn't an actual "page".

I also had to workaround the state by persisting it into the local storage, because changing the item data will refresh the preview page url, thus removing the `?language=X` I added myself to the preview window.

Would be nice if these extensions could also post messages to the target iframe to allow for custom interactions with the preview window without triggering a full page reload. For example, a "theme" switcher, where the preview window posts a message to the iframe indicating that theme changed, and the page will listen and react to that by changing a class in the body.

### Detailed Design

TBD

### Must Have:

- Additional (custom) configuration interfaces for the preview window, like the device change, refresh button and open external.
- Variables from those interfaces that can be inject into the preview URL.

### Could Have:

- A way to write custom interface elements for the preview window, and output values that can be added to the final preview url
- The custom interface must have access to the current preview state
  - The collection, item and fields data
  - State from other elements/components of the preview window (for example a "dialect" dropdown must have access to the "language" value)

### Drawbacks

It needs to be properly documented just like any other extension type (but TBH, IMHO this could be part of the implementation)

### Alternatives

I'm actually writing an extra page to use as a preview, so that I can toggle those options. The problem with this is that for example, the device button is completely useless since it will change my custom interface size too, which makes me remove the builtin device button with CSS and implement my own resize behavior in the wrapper itself.

### Adoption Strategy

It's just an additional feature

### Unresolved Questions

How the extension or the configured elements would look like?
Can we configure which of these interface elements are available/shown by default for all collections?
If it's an extension, how can I toggle which ones to be used in a specific collection?

**Comment by rijkvanzanten** [comment]
Heya! Thanks for opening this feature request!

This feature request has received over **15 votes** from the community. This means we'll move this feature request to the Under Review state! 

The Core team will schedule a meeting to review this request as soon as possible. The discussion will then be approved or denied. You may or may not be invited to join this meeting with the core team.

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
